multiraft/multiraft.go

```diff
@@ -428,14 +436,18 @@ type state struct {
 	groups    map[proto.RangeID]*group
 	nodes     map[proto.RaftNodeID]*node
 	writeTask *writeTask
+	// buffer the event if s.Events channel is full to avoid the deadlock
+	// between s.Events channel and callbackChan.
+	pendingEvents []interface{}
```

Comment:
I am not sure if we keep the buffer in the channel can help for performance as the channel filling will not depend on the select loop, so the event consuming goroutine in can keep busy if there is any event in the buffering channel.